### ISSUE TITLE
Feature for nginx location aliased to a persistent volume

### ISSUE THREAD

**Issue by johncronan** [caprover/caprover#1343]
**Is your feature request related to a problem? Please describe.**
I've adapted the docker-compose.yml file that I wrote for [Formative](https://github.com/johncronan/formative), in the hopes of making this a CapRover one-click app. This is a form builder app that handles media files similar to a CMS - user uploads might include large video files.

And it works! I came up with something that looks like this:

```
captainVersion: 4
services:
    $$cap_appname-db:
        image: postgres:13
        [...]
    $$cap_appname-mta:
        image: juanluisbaptiste/postfix:1.3
        [...]
    $$cap_appname:
        image: formative/formative:latest
        depends_on:
            - $$cap_appname-db
            - $$cap_appname-mta
        environment:
            [lots of environment variables ...]
        caproverExtra:
            containerHttpPort: ‘8000’
        volumes:
            - $$cap_appname-plugins:/opt/services/djangoapp/requirements
            - /captain/data/nginx-shared/formative_media:/opt/services/djangoapp/media
caproverOneClickApp:
    variables:
        - id: $$cap_formative_admin
          label: Admin username
          defaultValue: admin
          description:
          validRegex: /^([a-zA-Z0-9_@\.-])+$/
        - id: $$cap_formative_email
          label: Admin email
          description:
          validRegex: /^([a-zA-Z0-9_@\.-])+$/
        - id: $$cap_formative_password
          label: Admin password
          description:
          validRegex: /.{1,}/
        - id: $$cap_db_username
          label: Database username
          description:
          validRegex: /^([a-zA-Z0-9_])+$/
        - id: $$cap_db_password
          label: Database password
          description:
          validRegex: /.{1,}/
        - id: $$cap_db_database
          label: Database name
          defaultValue: formative
          description:
          validRegex: /^([a-zA-Z0-9_])+$/
          [some more ...]
    displayName: Formative
    isOfficial: false
    description: Self-hosted web app for collecting form responses and files. For application forms, you can use it together with reviewpanel to review and score applicant submissions.
```

The thing is, I've had to resort to creating a new directory under nginx-shared for the media volume. Eventually, I want to have access control enforced by my Django application, using X-Accel-Redirect to serve the files from nginx. Routing all that data transfer through gunicorn would compromise the performance pretty significantly.

Therefore, following the deployment we still need to add a `location /media/ { alias /nginx-shared/formative_media/; }` block to the nginx config.

I've also noted the following 2 problems:

- my containerHttpPort directive isn't working - I have to go to the app after it deploys and set the port to 8000. should I open a separate issue for this, or is it something simple that I've missed?
- since 'command' is not among the supported docker-compose fields, there is no way for me to use the official postgres image and apply configuration changes (even after moving them from command line arguments to a config file, still it is necessary to add `-c configfilelocation.conf` to the arguments). so same question - should I perhaps open a separate issue?

**Describe the solution you'd like**
With regard to the media volume issue, in my regular docker-compose.yml file, I create a docker volume that is mounted to both containers, nginx and the app.

Here's the solution that I would most prefer: a new field in caproverExtra that allows specifying a mapping between one of the service's volumes and an alias location that gets added to the nginx config. That would require adding to the nginx container's volumes, leaving nginx-shared out of it.

However, my goal is simply this: remain self-hosted, and achieve one-click deployment. I'd like to find out what you all think.

**Describe alternatives you've considered**
The only alternative that comes to mind, given that I want to remain fully self-hosted, is to require something else in addition to CapRover as a dependency for installation. That is, to run an additional server at the host level that handles serving the media. I could make my app one click to deploy, but then it would need to prompt the admin to configure this additional capability. I haven't thought deeply about what that would mean. It seems not ideal.

Thanks very much for CapRover!

**Comment by johncronan** [comment]
I want to be more clear about the "solution I'd like" part.

I think what's most needed is support for the X-Accel-Redirect scenario, so that means a caproverExtra field that creates an nginx alias location that's *internal*.

```
location /specified-redir-location {
    internal;
    alias /path/to/volume/for/nginx;
}
```

By narrowing the scope I hope that makes it more appealing, knowing that there's complications to deal with adding an option like this.

**Comment by githubsaturn** [comment]
> my containerHttpPort directive isn't working - I have to go to the app after it deploys and set the port to 8000. should I open a separate issue for this, or is it something simple that I've missed?

Lots of apps are using this, [example here](https://github.com/caprover/one-click-apps/blob/a8b0ddc059767aaa6ffbca4abedecc67439896a7/public/v4/apps/adminer.yml#L9-L10). Your quotations are bakc-quotes. `’` is different from `'`


> since 'command' is not among the supported docker-compose fields, there is no way for me to use the official postgres image and apply configuration changes (even after moving them from command line arguments to a config file, still it is necessary to add -c configfilelocation.conf to the arguments). so same question - should I perhaps open a separate issue?

Open a feature request for command. But there is already an alternative solution. [Like this](https://github.com/caprover/one-click-apps/blob/a8b0ddc059767aaa6ffbca4abedecc67439896a7/public/v4/apps/redis.yml#L10-L12).
 

> Here's the solution that I would most prefer: a new field in caproverExtra that allows specifying a mapping between one of the service's volumes and an alias location that gets added to the nginx config. That would require adding to the nginx container's volumes, leaving nginx-shared out of it.

This is a hard no as it's an anti pattern. `captain-nginx` (the main nginx container) is not supposed to serve any static files. In simple use case, the apps might reside on the main node, but they might also be deployed on other nodes. Assuming that the directory exists for both the app and the nginx container is fundamentally flawed. 

The proper solution for these use cases is to have a second nginx in your app's compose file that takes care of re-routing and severing static files, visually:
```
                                                            ┌───────────────┐
                                                            │               │
                                                   ┌───────►│   app-python  │
                                                   │        │               │
                                  ┌────────────────┤        └───────────────┘
┌─────────────────┐               │                │
│  captain-nginx  ├──────────────►│    app-nginx   │
└─────────────────┘               │                │        ┌────────────────┐
                                  └────────────────┤        │                │
                                                   │        │  static files  │
                                                   └───────►│                │
                                                            └────────────────┘
```

The overhead for the secondary nginx is very minimal.

**Comment by johncronan** [comment]
`The overhead for the secondary nginx is very minimal.`

I will use this solution. That makes sense.

Thanks for the detailed response, and for spotting those quotes!

**Comment by johncronan** [comment]
I have now implemented this change.

@githubsaturn, can I ask about another detail? I know about `client_max_body_size` in the nginx configuration limiting the maximum size of file uploads that I implement. I'm working now on batch downloading of files: a URL that responds with a large zip file. Am I correct that the download size will be limited by the value of `proxy_max_temp_file_size` in that same configuration?

Is there any way to make how large a download can be not dependent on available disk space? This is an interface for admin users, so there's not really any denial-of-service concern, and I wouldn't mind dedicating a process wherever that might be needed.

**Comment by githubsaturn** [comment]
These are nginx questions, and not CapRover specific. It's better to ask these questions in an nginx forum.

But I **think** `proxy_max_temp_file_size`doesn't limit the download size, it just limits the temporary file size. You can try it out. You can specifically set it to zero to remove it. 

`X-Accel-Buffering` : I don't know what it is and how it works.

**Comment by johncronan** [comment]
Okay, I did some testing. If I download from the container running locally (it's a composer nginx, but matching what CapRover does), I can pull down a 2GB download no problem. But then I set up a custom network throttling profile in my browser to download at a little under half the prior speed. With this test, I get the message about "an upstream response is buffered to a temporary file" and the download ends up truncating at a little over 1GB, the default value for `proxy_max_temp_file_size`.

I did try the `X-Accel-Buffering` thing, but I better understand that, now - it only applies in the sendfile scenario, so that's not for CapRover. I'm not sure if there's even a way CapRover's nginx could be configured to buffer or not, depending on a header in the proxied response. If I come up with something, I guess I'll open a new issue for discussion of that.

Meantime, this limitation will not really present an issue, I just wanted to understand the relevant settings and what install instructions I should have. Kind of too bad, tho, because the most likely time for an admin to do a large download of stored files is going to be when the disk space is low.
